### PR TITLE
[WIP] Updates to use new etcd client library

### PR DESCRIPTION
**Pull request by timothysc**
This is primarily meant to make sure upstream is ok with some of the changes being made and initial feedback prior to the rest of the cleanup.  (Pipelining) 

The tests are still broken atm and I'm well aware.  

@wojtek-t for the review.

xref: #11962

### CODE REVIEW

**Comment by davidopp** [comment]
@wojtek-t is out until Thu - is that OK? If so, please reassign to him (it got auto-assigned to me). Thanks!

**Comment by timothysc** [comment]
@davidopp totally ok to wait, it's still a WIP.  Also, I do not have acls to change "Assignee".

File: pkg/storage/etcd/etcd_watcher.go [hunk 1]
```diff
@@ -117,26 +118,56 @@ func newEtcdWatcher(list bool, include includeFunc, filter storage.FilterFunc, e
...
 // etcdGetInitialWatchState turns an etcd Get request into a watch equivalent
-func etcdGetInitialWatchState(client tools.EtcdClient, key string, recursive bool, incoming chan<- *etcd.Response) (resourceVersion uint64, err error) {
-	resp, err := client.Get(key, false, recursive)
+func etcdGetInitialWatchState(ctx context.Context, client etcd.Client, key string, recursive bool, incoming chan<- *etcd.Response) (resourceVersion uint64, err error) {
+
+	keysAPI := etcd.NewKeysAPI(client)
+	opts := etcd.GetOptions{
+		Recursive: recursive,
+		Sort:      false,
+		Quorum:    true,
```
Comment (yichengq) [inline pkg/storage/etcd/etcd_watcher.go]: no need for Quorum IMO

File: pkg/storage/etcd/etcd_helper.go [hunk 2]
```diff
@@ -435,7 +482,14 @@ func (h *etcdHelper) GuaranteedUpdate(key string, ptrToType runtime.Object, igno
 
 		startTime := time.Now()
 		// Swap origBody with data, if origBody is the latest etcd data.
-		response, err := h.client.CompareAndSwap(key, string(data), ttl, origBody, index)
+		opts := etcd.SetOptions{
+			PrevValue: origBody,
+			PrevIndex: index,
+			PrevExist: "PrevExist",
```
Comment (yichengq) [inline pkg/storage/etcd/etcd_helper.go]: PrevExist: client.PrevExist

File: pkg/storage/etcd/etcd_watcher.go [hunk 3]
```diff
@@ -117,26 +118,56 @@ func newEtcdWatcher(list bool, include includeFunc, filter storage.FilterFunc, e
...
+
+	opts := etcd.WatcherOptions{
+		AfterIndex: resourceVersion,
+		Recursive:  w.list,
+	}
+
+	watcher := keyAPI.Watcher(key, &opts)
+	w.ctx, w.cancel = context.WithCancel(ctx)
+
+	resp, err := watcher.Next(w.ctx)
```
Comment (yichengq) [inline pkg/storage/etcd/etcd_watcher.go]: The Next only returns one watch result. Is it fine in this case?

File: pkg/storage/etcd/etcd_helper.go [hunk 4]
```diff
@@ -345,8 +381,14 @@ func (h *etcdHelper) List(key string, listObj runtime.Object) error {
...
 }
 
-func (h *etcdHelper) listEtcdNode(key string) ([]*etcd.Node, uint64, error) {
-	result, err := h.client.Get(key, true, true)
+func (h *etcdHelper) listEtcdNode(ctx context.Context, key string) ([]*etcd.Node, uint64, error) {
+	keysAPI := etcd.NewKeysAPI(h.client)
+	opts := etcd.GetOptions{
+		Recursive: true,
+		Sort:      true,
+		Quorum:    true,
```
Comment (yichengq) [inline pkg/storage/etcd/etcd_helper.go]: there is no need to use `Quorum` here based on original code.

File: pkg/storage/etcd/etcd_helper.go [hunk 5]
```diff
@@ -101,7 +119,14 @@ func (h *etcdHelper) Create(key string, obj, out runtime.Object, ttl uint64) err
...
 	startTime := time.Now()
-	response, err := h.client.Create(key, string(data), ttl)
+
+	// TODO: With new api
+	keysAPI := etcd.NewKeysAPI(h.client)
+	opts := etcd.CreateInOrderOptions{
+		TTL: time.Duration(ttl),
+	}
+
+	response, err := keysAPI.CreateInOrder(ctx, key, string(data), &opts)
```
Comment (yichengq) [inline pkg/storage/etcd/etcd_helper.go]: It should use `Set(ctx context.Context, key, value string, opts *SetOptions) (*Response, error)` and set `PrevNoExist`. CreateInOrder is used to create in-order keys in one directory and user cannot specify key name.

File: pkg/storage/etcd/etcd_helper.go [hunk 6]
```diff
@@ -73,22 +73,40 @@ func init() {
...
+	}
+
+	memberLength := len(members)
+	if 0 == memberLength {
+		return nil
+	}
+
+	mlist := make([]string, memberLength)
+	for idx, member := range members {
+		mlist[idx] = member.Name
```
Comment (yichengq) [inline pkg/storage/etcd/etcd_helper.go]: `GetCluster` should list ClientURLs instead of name.

File: pkg/storage/etcd/etcd_helper.go [hunk 7]
```diff
@@ -73,22 +73,40 @@ func init() {
 }
 
 // Codec provides access to the underlying codec being used by the implementation.
-func (h *etcdHelper) Codec() runtime.Codec {
+func (h *etcdHelper) Codec(ctx context.Context) runtime.Codec {
```
Comment (yichengq) [inline pkg/storage/etcd/etcd_helper.go]: I think this ctx parameter is not necessary.

**Comment by ncdc** [comment]
@kubernetes/rh-cluster-infra @smarterclayton

**Comment by wojtek-t** [comment]
I'm back from vacation - will take a look later today.

**Comment by timothysc** [comment]
So I'll cleanup the rebase in the a.m. Github seems to loose it on a forced branch overwrite.  

Either way, here are my notes: 
- Tests are green 
- IMHO the entire semantics around the inner workings of the clients is "suspect" 
- kube2sky integration (amazed it works, it has a spamming loop which I could see if it fails it would always fail). 
- tool.EtcdClient needs to die. 
- Test rely on hidden semantics of the fake client, and re-written fake storage client should be built up separately, and remove dependencies piecemeal. 
- This was far more complicated then I had hoped, but in hindsight I also could have approached it differently, but it was a good learning experience, and I feel like I have a good handle on the storage and etcd layers.

**Comment by timothysc** [comment]
rebased, integration tests require re-cleaning.

**Comment by yichengq** [comment]
I cannot read the code because github tells me that `Sorry, we could not display the changes to this file because there were too many other changes to display.` :(

**Comment by timothysc** [comment]
@yichengq you should be able to click on the individual files.

**Comment by timothysc** [comment]
Still have some integration test failures.  Perhaps we can quick-chat on Monday regarding state.

**Comment by wojtek-t** [comment]
@jonboulle FYI
Note, that we decided to split it - so maybe it will be better to look at https://github.com/kubernetes/kubernetes/issues/11962

**Comment by wojtek-t** [comment]
Most of the changes were already upstreamed in other PRs so I'm closing this one.

**Pull request closed without merging**